Запрос на менторство BJS2-27001

vestmen opened the pull request:
Реализовал:
- Отправить запрос на менторство
- Получить все запросы на менторство по фильтрам
- Принять запрос на менторство
- Отклонить запрос на менторство

vestmen committed:
Запрос на менторство BJS2-27001

Реализовал:
- Отправить запрос на менторство
- Получить все запросы на менторство по фильтрам
- Принять запрос на менторство
- Отклонить запрос на менторство

natalia-dch commented:
Вроде как пакеты/директории обычно пишутся в snake_case?
ref https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html

natalia-dch commented on src/test/java/school/faang/user_service/service/mentorshipRequest/MentorshipRequestServiceTest.java:
```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Test
+    void testRequestByIdNotFound() {
```
А почему у rejectRequest это не проверяешь?

IvanTsegelnyuk commented on src/main/java/school/faang/user_service/controller/mentorshipRequest/MentorshipRequestController.java:
```diff
@@ -0,0 +1,51 @@
+
+    @PutMapping("reject/{id}")
+    public void rejectRequest(@PathVariable long id, @RequestBody RejectionDto rejection) {
+        mentorshipRequestService.rejectRequest(id, rejection.getReason());
```
В задании написано "Внутри метода нужно вызвать метод rejectRequest(id, rejection) класса MentorshipRequestService". Ты передаешь просто одно поле String

vestmen replied:
RejectionDto содержит в себе только поле String reason, в сервисе я его в метод передаю и id, а в контроллере - id и RejectionDto

amilosh commented on src/main/java/school/faang/user_service/controller/mentorshipRequest/MentorshipRequestController.java:
```diff
@@ -0,0 +1,51 @@
+import java.util.List;
+
+@RestController
+@RequestMapping("mentorship_requests")
```
mentorship-requests

Вроде через дефисы говорилось писать.

amilosh commented on src/main/java/school/faang/user_service/controller/mentorshipRequest/MentorshipRequestController.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    @PutMapping("reject/{id}")
+    public void rejectRequest(@PathVariable long id, @RequestBody RejectionDto rejection) {
```
Наверное полное название дай для 
RejectionDto
а то для других сущностей тоже вероятно есть реджекшены
чтобы не путаться

amilosh commented on src/main/java/school/faang/user_service/mapper/mentorshipRequest/MentorshipRequestMapper.java:
```diff
@@ -0,0 +1,13 @@
+import school.faang.user_service.dto.mentorshipRequest.MentorshipRequestDto;
+import school.faang.user_service.entity.MentorshipRequest;
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = org.mapstruct.ReportingPolicy.IGNORE)
```
org.mapstruct.ReportingPolicy.IGNORE
static import

amilosh commented on src/main/java/school/faang/user_service/service/mentorshipRequest/MentorshipRequestService.java:
```diff
@@ -0,0 +1,79 @@
+    public void rejectRequest(long id, String reason) {
+        log.info("Find mentorship request with id {}", id);
+        MentorshipRequest mentorshipRequest = findMentorshipRequestById(id);
+        mentorshipRequest.setStatus(RequestStatus.REJECTED);
```
я на енамы статические импорты всегда использую

кроме случаев, когда два енама имеют одинаковой значение

vestmen replied:
Да, я в других местах так и сделал, тут пропустил, спасибо

amilosh commented on src/main/java/school/faang/user_service/service/mentorshipRequest/MentorshipRequestService.java:
```diff
@@ -0,0 +1,79 @@
+
+    private MentorshipRequest findMentorshipRequestById(long id) {
+        return mentorshipRequestRepository.findById(id)
+                .orElseThrow(() -> {
+                    log.error("Request with id {} not found", id);
+                    return new IllegalArgumentException(
+                            String.format(REQUEST_NOT_FOUND, id));
+                });
```
Я тут подход использую с ReourceNotFoundException
и логированием в @ControllerAdvice
можешь посмотреть в моём репозитории
может быть понравится

vestmen committed:
Запрос на менторство BJS2-27001

Изменил наименование package

vestmen committed:
Запрос на менторство BJS2-27001

Изменил наименование package

vestmen committed:
Запрос на менторство BJS2-27001

Добавил возврат сущности в MentorshipRequestRepository и вернул в контроллере DTO вместо самой сущности

vestmen committed:
Запрос на менторство BJS2-27001

Внес правки

20cognosce commented on src/main/java/school/faang/user_service/controller/mentorship_request/MentorshipRequestController.java:
```diff
@@ -0,0 +1,52 @@
+        return mentorshipRequestMapper.toDto(mentorshipRequest);
+    }
+
+    @GetMapping
```
Post, потому что у гет нет тела по спецификации open api

vestmen replied:
переделал

20cognosce commented on src/main/java/school/faang/user_service/controller/mentorship_request/MentorshipRequestController.java:
```diff
@@ -0,0 +1,52 @@
+    @PostMapping
+    public MentorshipRequestDto requestMentorship(@RequestBody MentorshipRequestDto mentorshipRequestDto) {
+        MentorshipRequest mentorshipRequest = mentorshipRequestService.requestMentorship(
+                mentorshipRequestDto.getRequesterId(),
+                mentorshipRequestDto.getReceiverId(),
+                mentorshipRequestDto.getDescription());
```
Почему только 3 поля, для кого в ДТО есть все остальные?

vestmen replied:
Для создания запроса пользователь передает только 3 параметра, статус и время генерируются сами, а для запроса по фильтрам пользователю необходимо отдать все параметры запроса, включая статус и время

20cognosce commented on src/main/java/school/faang/user_service/service/mentorship_request/MentorshipRequestParametersChecker.java:
```diff
@@ -0,0 +1,91 @@
+
+    public void checkExistAcceptedRequest(long requesterId, long receiverId) {
+        log.info("Check exist accepted request");
+        if (mentorshipRequestRepository.existAcceptedRequest(requesterId, receiverId)) {
```
Так внутри if не делать, переменную объяви

vestmen replied:
переделал

20cognosce commented on src/main/java/school/faang/user_service/service/mentorship_request/MentorshipRequestParametersChecker.java:
```diff
@@ -0,0 +1,91 @@
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class MentorshipRequestParametersChecker {
```
Helper
не Validator как ты его назвал в service

vestmen replied:
исправил

20cognosce commented on src/main/java/school/faang/user_service/service/mentorship_request/MentorshipRequestService.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Transactional
+    public MentorshipRequest requestMentorship(long requesterId, long receiverId, String description) {
+        log.info("Start validation of params");
+        validator.checkRequestParams(requesterId, receiverId, description);
+        log.info("Validation of params successful");
+        MentorshipRequest mentorshipRequest = mentorshipRequestRepository.create(requesterId, receiverId, description);
+        log.info("Mentorship request from user with id {} to user with id {} created", requesterId, receiverId);
```
Бро, много логов
Реально перебор...

Не логируй каждую строчку, и в checker тоже
Только нужное

vestmen replied:
исправил

20cognosce commented on src/main/java/school/faang/user_service/service/mentorship_request/MentorshipRequestService.java:
```diff
@@ -0,0 +1,80 @@
+
+    @Transactional
+    public void acceptRequest(long id) {
+        MentorshipRequest mentorshipRequest = findMentorshipRequestById(id);
+        log.info("Request with id {} was found", id);
+        User requester = mentorshipRequest.getRequester();
+        User receiver = mentorshipRequest.getReceiver();
+        validator.checkExistAcceptedRequest(requester.getId(), receiver.getId());
+        log.info("Accept request from user with id {} to user with id {} not found", requester.getId(), receiver.getId());
+        requester.getMentors().add(receiver);
+        log.info("User with id {} added to the list of mentors by user with id {}", receiver.getId(), requester.getId());
+        mentorshipRequest.setStatus(ACCEPTED);
+        log.info("Request with id {} accepted", id);
```
Жесть с логами

vestmen replied:
убрал часть

20cognosce commented on src/main/java/school/faang/user_service/service/mentorship_request/MentorshipRequestService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @Transactional
+    public void acceptRequest(long id) {
```
А ты проверил работу этого метода?
Ты изменил
requester.getMentors().add(receiver);
mentorshipRequest.setStatus(ACCEPTED); 

Но в базу нифига не сохранил

20cognosce replied:
вместо void возвращай результат из БД

vestmen replied:
вернул запрос, и использовал метод save, но я проверял через Postman запрос меняется и без его явного сохранения в БД, так же и у юзера появляется связь меторства

20cognosce commented on src/main/java/school/faang/user_service/service/mentorship_request/MentorshipRequestService.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    @Transactional
+    public void rejectRequest(long id, String reason) {
```
Аналогично, ты ничего не сделал

20cognosce commented on src/main/java/school/faang/user_service/service/mentorship_request/error_messages/MentorshipRequestErrorMessages.java:
```diff
@@ -0,0 +1,11 @@
+package school.faang.user_service.service.mentorship_request.error_messages;
+
+public class MentorshipRequestErrorMessages {
```
@UtilityClass

vestmen replied:
переделал

20cognosce commented on src/test/java/school/faang/user_service/mapper/mentorship_request/MentorshipRequestMapperTest.java:
```diff
@@ -0,0 +1,50 @@
+        MentorshipRequestDto mentorshipRequestDto = mentorshipRequestMapper.toDto(mentorshipRequest);
+
+        assertNotNull(mentorshipRequestDto);
+        assertEquals(mentorshipRequestDto.getDescription(), mentorshipRequest.getDescription());
+        assertEquals(mentorshipRequestDto.getRequesterId(), mentorshipRequest.getRequester().getId());
+        assertEquals(mentorshipRequestDto.getReceiverId(), mentorshipRequest.getReceiver().getId());
+        assertEquals(mentorshipRequestDto.getStatus(), mentorshipRequest.getStatus());
+        assertEquals(mentorshipRequestDto.getRejectionReason(), mentorshipRequest.getRejectionReason());
+        assertEquals(mentorshipRequestDto.getCreatedAt(), mentorshipRequest.getCreatedAt());
+        assertEquals(mentorshipRequestDto.getUpdatedAt(), mentorshipRequest.getUpdatedAt());
```
AssertThat()
recursive comparison

vestmen replied:
Переделал

20cognosce reviewed (changes requested):
Вполне чистый код

20cognosce commented on src/main/java/school/faang/user_service/service/mentorship_request/MentorshipRequestParametersChecker.java:
```diff
@@ -0,0 +1,91 @@
+        MentorshipRequest latestMentorshipRequest = mentorshipRequestRepository
+                .findLatestRequest(requesterId, receiverId)
+                .orElse(null);
+        if (latestMentorshipRequest == null) {
+            log.info("This is first request from user with id {} to user with id {}",
+                    requesterId, receiverId);
+        } else {
+            log.info("Latest request from user with id {} to user with id {} was {}",
+                    requesterId, receiverId, latestMentorshipRequest.getCreatedAt());
+        }
```
if else для лога - это очень неоправданно

vestmen replied:
Переделал

20cognosce commented on src/test/java/school/faang/user_service/service/mentorship_request/MentorshipRequestServiceTest.java:
```diff
@@ -0,0 +1,194 @@
+
+        mentorshipRequestService.rejectRequest(mentorshipRequestId, reason);
+
+        assertEquals(mentorshipRequest.getStatus(), REJECTED);
+        assertEquals(mentorshipRequest.getRejectionReason(), reason);
```
проверяй что в базу сохраняется обновленная сущность

vestmen replied:
переделал

vestmen committed:
BJS2-27001 внес правки по замечаниям после ревью

20cognosce commented on src/test/java/school/faang/user_service/mapper/mentorship_request/MentorshipRequestMapperTest.java:
```diff
@@ -0,0 +1,47 @@
+        user1.setId(1L);
+        User user2 = new User();
+        user2.setId(2L);
+        MentorshipRequest mentorshipRequest = new MentorshipRequest();
+        mentorshipRequest.setRequester(user1);
+        mentorshipRequest.setReceiver(user2);
+        mentorshipRequest.setDescription("description");
+        mentorshipRequest.setStatus(RequestStatus.REJECTED);
+        mentorshipRequest.setRejectionReason("RejectionReason");
+        mentorshipRequest.setCreatedAt(LocalDateTime.now().minusDays(1));
+        mentorshipRequest.setUpdatedAt(LocalDateTime.now());
+        MentorshipRequestDto mentorshipRequestDto = mentorshipRequestMapper.toDto(mentorshipRequest);
```
Про Builder не забывай

20cognosce commented on src/test/java/school/faang/user_service/service/mentorship_request/MentorshipRequestServiceTest.java:
```diff
@@ -0,0 +1,206 @@
+        assertEquals(mentorshipRequest.getRejectionReason(), reason);
+    }
+
+    private void verifySaveRequest(MentorshipRequest mentorshipRequest) {
```
Зачем плодить отдельный метод на одну строчку?
